pynicotine/shares.py

```diff
@@ -112,50 +80,192 @@ class FileTypes:
+
+    def __setitem__(self, key, value):
+
+        encoded_key = str(key).encode("utf-8")
+        pickled_value = dumps(value, protocol=HIGHEST_PROTOCOL)
+
+        key_length = len(encoded_key)
+        length_data = self.PACK_LENGTHS(key_length, len(pickled_value))
+        item_data = (length_data + encoded_key + pickled_value)
+
+        self._file_handle.write(item_data)
+
+        self._value_offsets[key] = self._file_offset + self.LENGTH_DATA_SIZE + key_length
+        self._file_offset += len(item_data)
+
+    def get(self, key, default=None):
+
+        if key in self._value_offsets:
+            return self[key]
+
+        return default
+
+    def update(self, obj):
+        for key, value in obj.items():
+            self[key] = value
+
+    def close(self):
+
+        if self._overwrite:
+            os.fsync(self._file_handle)
+
+        self._file_handle.close()
+
+
 class Scanner:
     """Separate process responsible for building shares.
 
     It handles scanning of folders and files, as well as building
     databases and writing them to disk.
     """
 
-    def __init__(self, config_obj, queue, shared_folders, share_db_paths, init=False, rescan=True, rebuild=False):
+    def __init__(self, config_obj, queue, shared_folders, share_db_paths, init=False, rescan=True,
+                 rebuild=False, is_buddy_visible=False, is_trusted_visible=False):
 
         self.config = config_obj
         self.queue = queue
-        self.shared_folders, self.shared_buddy_folders = shared_folders
+        self.shared_folders, self.shared_buddy_folders, self.shared_trusted_folders = shared_folders
         self.share_dbs = {}
         self.share_db_paths = share_db_paths
         self.init = init
         self.rescan = rescan
         self.rebuild = rebuild
+        self.is_buddy_visible = is_buddy_visible
+        self.is_trusted_visible = is_trusted_visible
         self.tinytag = None
-        self.version = 2
+        self.version = 3
 
     def run(self):
 
         try:
             rename_process(b"nicotine-scan")
 
-            if not Shares.load_shares(self.share_dbs, self.share_db_paths, remove_failed=True):
-                # Failed to load shares, rebuild
+            shares_loaded = Shares.load_shares(self.share_dbs, self.share_db_paths, remove_failed=True)
+            old_mtimes = self.share_dbs.get("mtimes")
+            share_version = None
+
+            if old_mtimes is not None:
+                share_version = old_mtimes.get("__NICOTINE_SHARE_VERSION__")
+
+            if not shares_loaded or share_version != self.version:
+                # Failed to load shares or version is invalid, rebuild
                 self.rescan = self.rebuild = True
 
             if self.init:
                 self.create_compressed_shares()
 
             if self.rescan:
-                start_num_folders = len(self.share_dbs.get("buddyfiles", {}))
+                start_num_folders = len(self.share_dbs.get("buddystreams", {}))
```

Comment:
`"buddystreams"` -> `"trustedstreams"`

Surely, it is needed to count the highest possible permission level?

Reply:
This isn't correct anymore, it should be a combination of all three levels, since they don't duplicate files/folders anymore.